what4/src/What4/Expr/App.hs

```diff
@@ -0,0 +1,1861 @@
+-- values, but the constructor is kept hidden. The preferred way to
+-- construct a 'ExprBoundVar' is to use 'freshBoundVar'.
+data ExprBoundVar t (tp :: BaseType) =
+  BVar { bvarId  :: {-# UNPACK #-} !(Nonce t tp)
+       , bvarLoc :: !ProgramLoc
+       , bvarName :: !SolverSymbol
+       , bvarType :: !(BaseTypeRepr tp)
+       , bvarKind :: !VarKind
+       , bvarAbstractValue :: !(Maybe (AbstractValue tp))
+       }
+
+instance Eq (ExprBoundVar t tp) where
+  x == y = bvarId x == bvarId y
+
+instance TestEquality (ExprBoundVar t) where
+  testEquality x y = testEquality (bvarId x) (bvarId y)
+
+instance Ord (ExprBoundVar t tp) where
+  compare x y = compare (bvarId x) (bvarId y)
+
+instance OrdF (ExprBoundVar t) where
+  compareF x y = compareF (bvarId x) (bvarId y)
+
+instance Hashable (ExprBoundVar t tp) where
+  hashWithSalt s x = hashWithSalt s (bvarId x)
+
+instance HashableF (ExprBoundVar t) where
+  hashWithSaltF = hashWithSalt
+
+------------------------------------------------------------------------
+-- NonceApp
+
+-- | Type @NonceApp t e tp@ encodes the top-level application of an
+-- 'Expr'. It includes expression forms that bind variables (contrast
+-- with 'App').
+--
+-- Parameter @t@ is a phantom type brand used to track nonces.
+-- Parameter @e@ is used everywhere a recursive sub-expression would
+-- go. Uses of the 'NonceApp' type will tie the knot through this
+-- parameter. Parameter @tp@ indicates the type of the expression.
+data NonceApp t (e :: BaseType -> Type) (tp :: BaseType) where
+  Annotation ::
+    !(BaseTypeRepr tp) ->
+    !(Nonce t tp) ->
+    !(e tp) ->
+    NonceApp t e tp
+
+  Forall :: !(ExprBoundVar t tp)
+         -> !(e BaseBoolType)
+         -> NonceApp t e BaseBoolType
+  Exists :: !(ExprBoundVar t tp)
+         -> !(e BaseBoolType)
+         -> NonceApp t e BaseBoolType
+
+  -- Create an array from a function
+  ArrayFromFn :: !(ExprSymFn t e (idx ::> itp) ret)
+              -> NonceApp t e (BaseArrayType (idx ::> itp) ret)
+
+  -- Create an array by mapping over one or more existing arrays.
+  MapOverArrays :: !(ExprSymFn t e (ctx::>d) r)
+                -> !(Ctx.Assignment BaseTypeRepr (idx ::> itp))
+                -> !(Ctx.Assignment (ArrayResultWrapper e (idx ::> itp)) (ctx::>d))
+                -> NonceApp t e (BaseArrayType (idx ::> itp) r)
+
+  -- This returns true if all the indices satisfying the given predicate equal true.
+  ArrayTrueOnEntries
+    :: !(ExprSymFn t e (idx ::> itp) BaseBoolType)
+    -> !(e (BaseArrayType (idx ::> itp) BaseBoolType))
+    -> NonceApp t e BaseBoolType
+
+  -- Apply a function to some arguments
+  FnApp :: !(ExprSymFn t e args ret)
+        -> !(Ctx.Assignment e args)
+        -> NonceApp t e ret
+
+
+------------------------------------------------------------------------
+-- ExprSymFn
+
+-- | This describes information about an undefined or defined function.
+-- Parameter @t@ is a phantom type brand used to track nonces.
```

Comment:
Parameter `e` is new but there's no description of its use.  Adding this would especially help downstream porting efforts.

Reply:
I see it described in the NonceApp above... presumably the same thing?